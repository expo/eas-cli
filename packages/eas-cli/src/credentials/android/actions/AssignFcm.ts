import {
  AndroidFcmFragment,
  CommonAndroidAppCredentialsFragment,
} from '../../../graphql/generated';
import Log from '../../../log';
import { CredentialsContext } from '../../context';
import { AppLookupParams } from '../api/GraphqlClient';

export class AssignFcm {
  constructor(private app: AppLookupParams) {}

  public async runAsync(
    ctx: CredentialsContext,
    fcm: AndroidFcmFragment
  ): Promise<CommonAndroidAppCredentialsFragment> {
    const appCredentials =
      await ctx.android.createOrGetExistingAndroidAppCredentialsWithBuildCredentialsAsync(
        ctx.graphqlClient,
        this.app
      );
    const updatedAppCredentials = await ctx.android.updateAndroidAppCredentialsAsync(
      ctx.graphqlClient,
      appCredentials,
      {
        androidFcmId: fcm.id,
      }
    );
    Log.succeed(`FCM API Key assigned to ${this.app.androidApplicationIdentifier}`);
    return updatedAppCredentials;
  }
}
