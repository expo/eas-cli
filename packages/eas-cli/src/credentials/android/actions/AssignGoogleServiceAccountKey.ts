import {
  CommonAndroidAppCredentialsFragment,
  GoogleServiceAccountKeyFragment,
} from '../../../graphql/generated';
import Log from '../../../log';
import { CredentialsContext } from '../../context';
import { AppLookupParams } from '../api/GraphqlClient';

export class AssignGoogleServiceAccountKey {
  constructor(private app: AppLookupParams) {}

  public async runAsync(
    ctx: CredentialsContext,
    googleServiceAccountKey: GoogleServiceAccountKeyFragment
  ): Promise<CommonAndroidAppCredentialsFragment> {
    const appCredentials =
      await ctx.android.createOrGetExistingAndroidAppCredentialsWithBuildCredentialsAsync(
        ctx.graphqlClient,
        this.app
      );
    const updatedAppCredentials = await ctx.android.updateAndroidAppCredentialsAsync(
      ctx.graphqlClient,
      appCredentials,
      {
        googleServiceAccountKeyForSubmissionsId: googleServiceAccountKey.id,
      }
    );
    Log.succeed(
      `Google Service Account Key assigned to ${this.app.androidApplicationIdentifier} for submissions`
    );
    return updatedAppCredentials;
  }
}
