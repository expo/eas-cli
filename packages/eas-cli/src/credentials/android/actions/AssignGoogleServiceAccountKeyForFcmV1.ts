import {
  CommonAndroidAppCredentialsFragment,
  GoogleServiceAccountKeyFragment,
} from '../../../graphql/generated';
import Log from '../../../log';
import { CredentialsContext } from '../../context';
import { AppLookupParams } from '../api/GraphqlClient';

export class AssignGoogleServiceAccountKeyForFcmV1 {
  constructor(private app: AppLookupParams) {}

  public async runAsync(
    ctx: CredentialsContext,
    googleServiceAccountKey: GoogleServiceAccountKeyFragment
  ): Promise<CommonAndroidAppCredentialsFragment> {
    const appCredentials =
      await ctx.android.createOrGetExistingAndroidAppCredentialsWithBuildCredentialsAsync(
        ctx.graphqlClient,
        this.app
      );
    const updatedAppCredentials = await ctx.android.updateAndroidAppCredentialsAsync(
      ctx.graphqlClient,
      appCredentials,
      {
        googleServiceAccountKeyForFcmV1Id: googleServiceAccountKey.id,
      }
    );
    Log.succeed(
      `Google Service Account Key assigned to ${this.app.androidApplicationIdentifier} for FCM V1`
    );
    return updatedAppCredentials;
  }
}
