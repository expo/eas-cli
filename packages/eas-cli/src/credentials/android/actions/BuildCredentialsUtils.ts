import assert from 'assert';
import { nanoid } from 'nanoid';

import { AndroidAppBuildCredentialsFragment } from '../../../graphql/generated';
import { ora } from '../../../ora';
import { getApplicationIdAsync } from '../../../project/android/applicationId';
import { GradleBuildContext } from '../../../project/android/gradle';
import { getProjectAccountName, getProjectConfigDescription } from '../../../project/projectUtils';
import { promptAsync } from '../../../prompts';
import { findAccountByName } from '../../../user/Account';
import { CredentialsContext } from '../../context';
import { AppLookupParams } from '../api/GraphqlClient';

/**
 * Legacy credentials can be copied over to EAS if the user does not have
 * EAS credentials set up yet
 */
export async function canCopyLegacyCredentialsAsync(
  ctx: CredentialsContext,
  app: AppLookupParams
): Promise<boolean> {
  const appCredentials = await ctx.android.getAndroidAppCredentialsWithCommonFieldsAsync(app);
  if (appCredentials) {
    return false; // modern credentials already exist
  }

  const legacyAppCredentials =
    await ctx.android.getLegacyAndroidAppCredentialsWithCommonFieldsAsync(app);
  return !!legacyAppCredentials; // user has some legacy credentials
}

export async function promptUserAndCopyLegacyCredentialsAsync(
  ctx: CredentialsContext,
  app: AppLookupParams
): Promise<void> {
  assert(
    await canCopyLegacyCredentialsAsync(ctx, app),
    'User not eligible to copy classic build credentials to EAS'
  );

  const spinner = ora('Classic credentials detected, copying to EAS...').start();

  try {
    const legacyAppCredentials =
      await ctx.android.getLegacyAndroidAppCredentialsWithCommonFieldsAsync(app);
    if (!legacyAppCredentials) {
      return;
    }

    const appCredentials =
      await ctx.android.createOrGetExistingAndroidAppCredentialsWithBuildCredentialsAsync(app);
    const legacyFcm = legacyAppCredentials.androidFcm;
    if (legacyFcm) {
      const clonedFcm = await ctx.android.createFcmAsync(
        app.account,
        legacyFcm.credential,
        legacyFcm.version
      );
      await ctx.android.updateAndroidAppCredentialsAsync(appCredentials, {
        androidFcmId: clonedFcm.id,
      });
    }

    const legacyBuildCredentials = await ctx.android.getLegacyAndroidAppBuildCredentialsAsync(app);
    const legacyKeystore = legacyBuildCredentials?.androidKeystore ?? null;

    if (legacyKeystore) {
      const clonedKeystore = await ctx.android.createKeystoreAsync(app.account, {
        keystore: legacyKeystore.keystore,
        keystorePassword: legacyKeystore.keystorePassword,
        keyAlias: legacyKeystore.keyAlias,
        keyPassword: legacyKeystore.keyPassword ?? undefined,
        type: legacyKeystore.type,
      });
      await createOrUpdateDefaultAndroidAppBuildCredentialsAsync(ctx, app, {
        androidKeystoreId: clonedKeystore.id,
      });
    }
  } catch (e) {
    spinner.fail(`Unable to migrate credentials to EAS.`);
    throw e;
  }

  spinner.succeed('Credentials copied to EAS.');
}

export async function getAppLookupParamsFromContextAsync(
  ctx: CredentialsContext,
  gradleContext?: GradleBuildContext
): Promise<AppLookupParams> {
  ctx.ensureProjectContext();
  const projectName = ctx.exp.slug;
  const accountName = getProjectAccountName(ctx.exp, ctx.user);
  const account = findAccountByName(ctx.user.accounts, accountName);
  if (!account) {
    throw new Error(`You do not have access to account: ${accountName}`);
  }

  const androidApplicationIdentifier = await getApplicationIdAsync(
    ctx.projectDir,
    ctx.exp,
    gradleContext
  );
  if (!androidApplicationIdentifier) {
    throw new Error(
      `android.package needs to be defined in your ${getProjectConfigDescription(
        ctx.projectDir
      )} file`
    );
  }

  return { account, projectName, androidApplicationIdentifier };
}

export async function createOrUpdateDefaultAndroidAppBuildCredentialsAsync(
  ctx: CredentialsContext,
  appLookupParams: AppLookupParams,
  {
    androidKeystoreId,
  }: {
    androidKeystoreId: string;
  }
): Promise<AndroidAppBuildCredentialsFragment> {
  assert(
    !ctx.nonInteractive,
    'createOrUpdateDefaultAndroidAppBuildCredentialsAsync must be run in interactive mode'
  );
  const existingDefaultBuildCredentials =
    await ctx.android.getDefaultAndroidAppBuildCredentialsAsync(appLookupParams);
  if (existingDefaultBuildCredentials) {
    return await ctx.android.updateAndroidAppBuildCredentialsAsync(
      existingDefaultBuildCredentials,
      { androidKeystoreId }
    );
  }
  return await ctx.android.createAndroidAppBuildCredentialsAsync(appLookupParams, {
    name: generateRandomName(),
    isDefault: true,
    androidKeystoreId,
  });
}

export async function promptForNameAsync(): Promise<string> {
  const { providedName } = await promptAsync({
    type: 'text',
    name: 'providedName',
    message: 'Assign a name to your build credentials:',
    initial: generateRandomName(),
    validate: (input: string) => input !== '',
  });
  return providedName;
}

/**
 * sort a build credentials array in descending order of preference
 * prefer default credentials, then prefer names that come first lexicographically
 */
export function sortBuildCredentials(
  androidAppBuildCredentialsList: AndroidAppBuildCredentialsFragment[]
): AndroidAppBuildCredentialsFragment[] {
  return androidAppBuildCredentialsList.sort((buildCredentialsA, buildCredentialsB) => {
    if (buildCredentialsA.isDefault) {
      return -1;
    } else if (buildCredentialsB.isDefault) {
      return 1;
    }
    return buildCredentialsA.name.localeCompare(buildCredentialsB.name);
  });
}

function generateRandomName(): string {
  return `Build Credentials ${nanoid(10)}`;
}
