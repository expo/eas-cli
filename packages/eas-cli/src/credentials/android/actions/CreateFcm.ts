import { AccountFragment, AndroidFcmFragment, AndroidFcmVersion } from '../../../graphql/generated';
import Log from '../../../log';
import { promptAsync } from '../../../prompts';
import { CredentialsContext } from '../../context';

export class CreateFcm {
  constructor(private account: AccountFragment) {}

  public async runAsync(ctx: CredentialsContext): Promise<AndroidFcmFragment> {
    if (ctx.nonInteractive) {
      throw new Error(`FCM API Key cannot be uploaded in non-interactive mode.`);
    }
    const { fcmApiKey } = await promptAsync([
      {
        type: 'text',
        name: 'fcmApiKey',
        message: 'FCM API Key',
        validate: (value: string) => value.length > 0 || "FCM API Key can't be empty",
      },
    ]);
    const fcmFragment = await ctx.android.createFcmAsync(
      ctx.graphqlClient,
      this.account,
      fcmApiKey,
      AndroidFcmVersion.Legacy
    );
    Log.succeed('Uploaded FCM API Key');
    return fcmFragment;
  }
}
