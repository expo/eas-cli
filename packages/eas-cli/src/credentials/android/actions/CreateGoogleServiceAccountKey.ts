import chalk from 'chalk';
import fs from 'fs-extra';

import { AccountFragment, GoogleServiceAccountKeyFragment } from '../../../graphql/generated';
import Log, { learnMore } from '../../../log';
import { promptAsync } from '../../../prompts';
import { CredentialsContext } from '../../context';
import { GoogleServiceAccountKey } from '../credentials';
import {
  detectGoogleServiceAccountKeyPathAsync,
  readAndValidateServiceAccountKey,
} from '../utils/googleServiceAccountKey';

export class CreateGoogleServiceAccountKey {
  constructor(private account: AccountFragment) {}

  public async runAsync(ctx: CredentialsContext): Promise<GoogleServiceAccountKeyFragment> {
    if (ctx.nonInteractive) {
      throw new Error(`New Google Service Account Key cannot be created in non-interactive mode.`);
    }
    const jsonKeyObject = await this.provideAsync(ctx);
    const gsaKeyFragment = await ctx.android.createGoogleServiceAccountKeyAsync(
      ctx.graphqlClient,
      this.account,
      jsonKeyObject
    );
    Log.succeed('Uploaded Google Service Account Key.');
    return gsaKeyFragment;
  }

  private async provideAsync(ctx: CredentialsContext): Promise<GoogleServiceAccountKey> {
    try {
      const keyJsonPath = await this.provideKeyJsonPathAsync(ctx);
      return readAndValidateServiceAccountKey(keyJsonPath);
    } catch (e) {
      Log.error(e);
      return await this.provideAsync(ctx);
    }
  }

  private async provideKeyJsonPathAsync(ctx: CredentialsContext): Promise<string> {
    const detectedPath = await detectGoogleServiceAccountKeyPathAsync(ctx.projectDir);
    if (detectedPath) {
      return detectedPath;
    }

    Log.log(
      `${chalk.bold(
        'A Google Service Account JSON key is required for uploading your app to Google Play Store, and for sending Android Notifications via FCM V1.'
      )}.\n` +
        `If you're not sure what this is or how to create one, ${learnMore(
          'https://expo.fyi/creating-google-service-account',
          { learnMoreMessage: 'learn more' }
        )}`
    );
    const { filePath } = await promptAsync({
      name: 'filePath',
      message: 'Path to Google Service Account file:',
      initial: 'api-0000000000000000000-111111-aaaaaabbbbbb.json',
      type: 'text',
      // eslint-disable-next-line async-protect/async-suffix
      validate: async (filePath: string) => {
        try {
          const stats = await fs.stat(filePath);
          if (stats.isFile()) {
            return true;
          }
          return 'Input is not a file.';
        } catch {
          return 'File does not exist.';
        }
      },
    });
    return filePath;
  }
}
