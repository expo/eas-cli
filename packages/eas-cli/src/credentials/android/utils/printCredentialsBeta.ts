import chalk from 'chalk';

import {
  AndroidAppBuildCredentialsFragment,
  AndroidFcmVersion,
  CommonAndroidAppCredentialsFragment,
  FcmSnippetLegacy,
  FcmSnippetV1,
} from '../../../graphql/generated';
import Log from '../../../log';
import { fromNow } from '../../../utils/date';
import { sortBuildCredentials } from '../actions/BuildCredentialsUtils';
import { AppLookupParams } from '../api/GraphqlClient';

export function displayEmptyAndroidCredentials(appLookupParams: AppLookupParams): void {
  const { projectName, androidApplicationIdentifier } = appLookupParams;
  Log.log(chalk.bold(`Android Credentials`));
  Log.log(`  Project: ${projectName}`);
  Log.log(`  Application Identifier: ${androidApplicationIdentifier}`);
  Log.log(`  No credentials set up yet!`);
}

function displayLegacyAndroidAppCredentials(
  legacyAppCredentials: CommonAndroidAppCredentialsFragment,
  appCredentials: CommonAndroidAppCredentialsFragment | null
): void {
  if (appCredentials) {
    // differentiate between old and new if there's both
    Log.log(chalk.bold(`  Credentials ported from Expo Classic (expo-cli):`));
  }
  displayAndroidFcmCredentials(legacyAppCredentials);
  // There can only be one set of build credentials in legacy app credentials
  const [legacyBuildCredentials] = legacyAppCredentials.androidAppBuildCredentialsList;
  if (legacyBuildCredentials) {
    displayAndroidBuildCredentials(
      legacyBuildCredentials,
      /* only one set to show, skip config display */ true
    );
  }
}

function displayAndroidFcmCredentials(appCredentials: CommonAndroidAppCredentialsFragment): void {
  const maybeFcm = appCredentials.androidFcm;
  Log.log(`  Push Notifications (FCM):`);
  if (!maybeFcm) {
    Log.log(`    None assigned yet`);
    Log.newLine();
    return;
  }
  const { snippet, version, updatedAt } = maybeFcm;
  if (version === AndroidFcmVersion.Legacy) {
    const { firstFourCharacters, lastFourCharacters } = snippet as FcmSnippetLegacy;
    Log.log(`    Key: ${firstFourCharacters}...${lastFourCharacters}`);
  } else if (version === AndroidFcmVersion.V1) {
    const { projectId, serviceAccountEmail, clientId, keyId } = snippet as FcmSnippetV1;
    Log.log(`    Project Id: ${projectId}`);
    Log.log(`    Service Account: ${serviceAccountEmail}`);
    Log.log(`    Client Id: ${clientId}`);
    Log.log(`    Key Id: ${keyId}`);
  }
  Log.log(`    Updated ${fromNow(new Date(updatedAt))} ago`);
  Log.newLine();
}

function displayEASAndroidAppCredentials(
  legacyAppCredentials: CommonAndroidAppCredentialsFragment | null,
  appCredentials: CommonAndroidAppCredentialsFragment
): void {
  if (legacyAppCredentials) {
    // differentiate between old and new if there's both
    Log.log(chalk.bold(`  EAS Credentials:`));
    Log.log(`  These will take precedence over Expo Classic credentials`);
    Log.newLine();
  }
  displayAndroidFcmCredentials(appCredentials);
  const sortedBuildCredentialsList = sortBuildCredentials(
    appCredentials.androidAppBuildCredentialsList
  );
  for (const buildCredentials of sortedBuildCredentialsList) {
    displayAndroidBuildCredentials(buildCredentials);
  }
}

function formatFingerprint(fingerprint: string | null): string {
  if (!fingerprint) {
    return 'unavailable';
  }
  const uppercaseFingerprint = fingerprint.toUpperCase();
  const bytes = [];
  for (let i = 0; i < uppercaseFingerprint.length; i++) {
    const halfByte = uppercaseFingerprint.charAt(i);
    if (i % 2 === 0) {
      bytes.push(halfByte); // first half of the byte
    } else {
      bytes[bytes.length - 1] += halfByte; // second half of the byte
    }
  }
  return bytes.join(':');
}
function displayAndroidBuildCredentials(
  buildCredentials: AndroidAppBuildCredentialsFragment,
  skipConfigurationDisplay?: boolean
): void {
  if (!skipConfigurationDisplay) {
    const { isDefault, name } = buildCredentials;
    Log.log(chalk.bold(`  Configuration: ${name}${isDefault ? ' (Default)' : ''}`));
  }

  const maybeKeystore = buildCredentials.androidKeystore;
  Log.log(`  Keystore:`);
  if (maybeKeystore) {
    const {
      keyAlias,
      type,
      md5CertificateFingerprint,
      sha1CertificateFingerprint,
      sha256CertificateFingerprint,
      updatedAt,
    } = maybeKeystore;
    Log.log(`    Type: ${type}`);
    Log.log(`    Key Alias: ${keyAlias}`);

    Log.log(`    MD5 Fingerprint: ${formatFingerprint(md5CertificateFingerprint ?? null)}`);
    Log.log(`    SHA1 Fingerprint: ${formatFingerprint(sha1CertificateFingerprint ?? null)}`);
    Log.log(`    SHA256 Fingerprint: ${formatFingerprint(sha256CertificateFingerprint ?? null)}`);
    Log.log(`    Updated ${fromNow(new Date(updatedAt))} ago`);
  } else {
    Log.log(`    None assigned yet`);
  }
  Log.newLine();
}

export function displayAndroidAppCredentials({
  appLookupParams,
  legacyAppCredentials,
  appCredentials,
}: {
  appLookupParams: AppLookupParams;
  legacyAppCredentials: CommonAndroidAppCredentialsFragment | null;
  appCredentials: CommonAndroidAppCredentialsFragment | null;
}): void {
  const { projectName, androidApplicationIdentifier } = appLookupParams;
  Log.log(chalk.bold(`Android Credentials`));
  Log.log(`  Project: ${projectName}`);
  Log.log(`  Application Identifier: ${androidApplicationIdentifier}`);
  Log.newLine();

  if (appCredentials) {
    displayEASAndroidAppCredentials(legacyAppCredentials, appCredentials);
  }

  if (legacyAppCredentials) {
    displayLegacyAndroidAppCredentials(legacyAppCredentials, appCredentials);
  }
}
