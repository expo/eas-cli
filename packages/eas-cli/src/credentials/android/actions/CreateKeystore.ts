import { AndroidKeystoreFragment } from '../../../graphql/generated';
import Log from '../../../log';
import { getProjectIdAsync } from '../../../project/projectUtils';
import { Account } from '../../../user/Account';
import { Context } from '../../context';
import { askForUserProvidedAsync } from '../../utils/promptForCredentials';
import { KeystoreWithType, keystoreSchema } from '../credentials';
import { generateRandomKeystoreAsync } from '../utils/keystore';
import { getKeystoreWithType, validateKeystore } from '../utils/keystoreNew';

export class CreateKeystore {
  constructor(private account: Account) {}

  public async runAsync(ctx: Context): Promise<AndroidKeystoreFragment> {
    if (ctx.nonInteractive) {
      throw new Error(`New keystore cannot be created in non-interactive mode.`);
    }

    const projectId = await getProjectIdAsync(ctx.exp);
    const keystore = await this.provideOrGenerateAsync(projectId);
    const keystoreFragment = await ctx.android.createKeystoreAsync(this.account, keystore);
    Log.succeed('Created keystore');
    return keystoreFragment;
  }

  private async provideOrGenerateAsync(projectId: string): Promise<KeystoreWithType> {
    const providedKeystore = await askForUserProvidedAsync(keystoreSchema);
    if (providedKeystore) {
      const providedKeystoreWithType = getKeystoreWithType(providedKeystore);
      validateKeystore(providedKeystoreWithType);
      return providedKeystoreWithType;
    }
    return await generateRandomKeystoreAsync(projectId);
  }
}
