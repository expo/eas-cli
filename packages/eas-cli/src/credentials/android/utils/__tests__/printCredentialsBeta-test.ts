import mockdate from 'mockdate';

import Log from '../../../../log';
import {
  testAndroidAppCredentialsFragment,
  testLegacyAndroidAppCredentialsFragment,
} from '../../../__tests__/fixtures-android-new';
import { createCtxMock } from '../../../__tests__/fixtures-context';
import { getAppLookupParamsFromContext } from '../../actions/BuildCredentialsUtils';
import { displayAndroidAppCredentials } from '../printCredentialsBeta';

jest.mock('../../../../log');
jest.mock('chalk', () => ({ bold: jest.fn(log => log) }));

mockdate.set(new Date('4/20/2021'));

describe('print credentials', () => {
  it('prints the AndroidAppCredentials fragment', async () => {
    const ctx = createCtxMock();
    const appLookupParams = getAppLookupParamsFromContext(ctx);
    displayAndroidAppCredentials({
      appLookupParams,
      legacyAppCredentials: testLegacyAndroidAppCredentialsFragment,
      appCredentials: testAndroidAppCredentialsFragment,
    });
    const loggedSoFar = (Log.log as jest.Mock).mock.calls.reduce(
      (acc, mockValue) => acc + mockValue
    );
    expect(loggedSoFar).toMatchSnapshot();
  });
});
