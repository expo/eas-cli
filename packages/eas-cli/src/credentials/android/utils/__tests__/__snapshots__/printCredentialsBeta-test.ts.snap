// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`print credentials prints the AndroidAppCredentials fragment 1`] = `"Android Credentials  Project: testApp  Application Identifier: test.com.app  Push Notifications (FCM):    Key: abcd...efgh    Updated 0 second ago  Configuration: legacy (Default)  Keystore:    Type: JKS    Key Alias: QHdrb3p5cmEvY3JlZGVudGlhbHMtdGVzdA==    MD5 Fingerprint: TE:ST:-M:D5    SHA1 Fingerprint: TE:ST:-S:HA:1    SHA256 Fingerprint: TE:ST:-S:HA:25:6    Updated 0 second ago"`;
